File: app/src/main/java/org/mozilla/focus/urlinput/UrlInputFragment.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.urlinput;

import android.app.Activity;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.mozilla.focus.R;
import org.mozilla.focus.autocomplete.UrlAutoCompleteFilter;
import org.mozilla.focus.home.HomeFragment;
import org.mozilla.focus.search.SearchEngineManager;
import org.mozilla.focus.telemetry.TelemetryWrapper;
import org.mozilla.focus.utils.SearchUtils;
import org.mozilla.focus.utils.SupportUtils;
import org.mozilla.focus.utils.ViewUtils;
import org.mozilla.focus.web.WebViewProvider;
import org.mozilla.focus.widget.FlowLayout;
import org.mozilla.focus.widget.FragmentListener;
import org.mozilla.focus.widget.InlineAutocompleteEditText;

import java.util.List;
import java.util.Locale;

/**
 * Fragment for displaying he URL input controls.
 */
public class UrlInputFragment extends Fragment implements UrlInputContract.View,
        View.OnClickListener, View.OnLongClickListener,
        InlineAutocompleteEditText.OnCommitListener, InlineAutocompleteEditText.OnFilterListener,
        InlineAutocompleteEditText.OnTextChangeListener {

    public static final String FRAGMENT_TAG = "url_input";

    private static final String ARGUMENT_URL = "url";
    private static final String ARGUMENT_PARENT_FRAGMENT = "parent_frag_tag";
    private static final String ARGUMENT_ALLOW_SUGGESTION = "allow_suggestion";

    private static final int REQUEST_THROTTLE_THRESHOLD = 300;

    private UrlInputContract.Presenter presenter;

    /**
     * Create a new UrlInputFragment and animate the url input view from the position/size of the
     * fake url bar view.
     */
    public static UrlInputFragment create(@Nullable String url, String parentFragmentTag, final boolean allowSuggestion) {
        Bundle arguments = new Bundle();
        arguments.putString(ARGUMENT_URL, url);
        arguments.putString(ARGUMENT_PARENT_FRAGMENT, parentFragmentTag);
        arguments.putBoolean(ARGUMENT_ALLOW_SUGGESTION, allowSuggestion);

        UrlInputFragment fragment = new UrlInputFragment();
        fragment.setArguments(arguments);

        return fragment;
    }

    private InlineAutocompleteEditText urlView;
    private FlowLayout suggestionView;
    private View clearView;

    private UrlAutoCompleteFilter urlAutoCompleteFilter;
    private boolean autoCompleteInProgress;
    private View dismissView;
    private long lastRequestTime;
    private boolean allowSuggestion;

    @Override
    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);
        final String userAgent = WebViewProvider.getUserAgentString(getActivity());
        this.presenter = new UrlInputPresenter(SearchEngineManager.getInstance()
                .getDefaultSearchEngine(getActivity()), userAgent);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_urlinput, container, false);

        dismissView = view.findViewById(R.id.dismiss);
        dismissView.setOnClickListener(this);

        clearView = view.findViewById(R.id.clear);
        clearView.setOnClickListener(this);

        urlAutoCompleteFilter = new UrlAutoCompleteFilter(getContext().getApplicationContext());

        suggestionView = (FlowLayout) view.findViewById(R.id.search_suggestion);

        urlView = (InlineAutocompleteEditText) view.findViewById(R.id.url_edit);
        urlView.setOnTextChangeListener(this);
        urlView.setOnFilterListener(this);
        urlView.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                // Avoid showing keyboard again when returning to the previous page by back key.
                if (hasFocus) {
                    ViewUtils.showKeyboard(urlView);
                } else {
                    ViewUtils.hideKeyboard(urlView);
                }
            }
        });

        urlView.setOnCommitListener(this);

        initByArguments();

        return view;
    }

    @Override
    public void onStart() {
        super.onStart();
        presenter.setView(this);
        urlView.requestFocus();
    }

    @Override
    public void onStop() {
        super.onStop();
        presenter.setView(null);
    }

    @Override
    public boolean onLongClick(View view) {
        switch (view.getId()) {
            case R.id.suggestion_item:
                setUrlText(((TextView) view).getText());
                TelemetryWrapper.searchSuggestionLongClick();
                return true;
            case R.id.clear:
                TelemetryWrapper.searchClear();
                return false;
            case R.id.dismiss:
                TelemetryWrapper.searchDismiss();
                return false;
            default:
                return false;
        }
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.clear:
                urlView.setText("");
                urlView.requestFocus();
                break;
            case R.id.dismiss:
                dismiss();
                break;
            case R.id.suggestion_item:
                onSuggestionClicked(((TextView) view).getText());
                break;
            default:
                throw new IllegalStateException("Unhandled view in onClick()");
        }
    }

    private void initByArguments() {
        final Bundle args = getArguments();
        if (args.containsKey(ARGUMENT_URL)) {
            final String url = args.getString(ARGUMENT_URL);
            urlView.setText(url);
            clearView.setVisibility(TextUtils.isEmpty(url) ? View.GONE : View.VISIBLE);
        }
        allowSuggestion = args.getBoolean(ARGUMENT_ALLOW_SUGGESTION, true);
    }

    private void onSuggestionClicked(CharSequence tag) {
        setUrlText(tag);
        onCommit(true);
    }

    private void dismiss() {
        // This method is called from animation callbacks. In the short time frame between the animation
        // starting and ending the activity can be paused. In this case this code can throw an
        // IllegalStateException because we already saved the state (of the activity / fragment) before
        // this transaction is committed. To avoid this we commit while allowing a state loss here.
        // We do not save any state in this fragment (It's getting destroyed) so this should not be a problem.
        final Activity activity = getActivity();
        if (activity instanceof FragmentListener) {
            ((FragmentListener) activity).onNotified(this, FragmentListener.TYPE.DISMISS_URL_INPUT, true);
        }
    }

    public void onCommit(boolean isSuggestion) {
        final String input = isSuggestion ? urlView.getOriginalText() : urlView.getText().toString();
        if (!input.trim().isEmpty()) {
            ViewUtils.hideKeyboard(urlView);

            final boolean isUrl = SupportUtils.isUrl(input);

            final String url = isUrl
                    ? SupportUtils.normalize(input)
                    : SearchUtils.createSearchUrl(getContext(), input);

            boolean isOpenInNewTab = openUrl(url);

            if (isOpenInNewTab) {
                TelemetryWrapper.addNewTabFromHome();
            }
            TelemetryWrapper.urlBarEvent(isUrl, isSuggestion);
        }
    }

    /**
     * @param url the URL to open
     * @return true if open URL in new tab.
     */
    private boolean openUrl(String url) {
        boolean openNewTab = false;

        Bundle args = getArguments();
        if (args != null && args.containsKey(ARGUMENT_PARENT_FRAGMENT)) {
            openNewTab = HomeFragment.FRAGMENT_TAG.equals(args.getString(ARGUMENT_PARENT_FRAGMENT));
        }

        final Activity activity = getActivity();
        if (activity instanceof FragmentListener) {
            final FragmentListener listener = (FragmentListener) activity;
            FragmentListener.TYPE msgType = openNewTab
                    ? FragmentListener.TYPE.OPEN_URL_IN_NEW_TAB
                    : FragmentListener.TYPE.OPEN_URL_IN_CURRENT_TAB;

            listener.onNotified(this, msgType, url);
        }

        return openNewTab;
    }

    @Override
    public void setUrlText(CharSequence text) {
        this.urlView.setOnTextChangeListener(null);
        this.urlView.setText(text);
        this.urlView.setSelection(text.length());
        this.urlView.setOnTextChangeListener(this);
    }

    @Override
    public void setSuggestions(@Nullable List<CharSequence> texts) {
        this.suggestionView.removeAllViews();
        if (texts == null) {
            return;
        }

        final String searchKey = urlView.getOriginalText().trim().toLowerCase(Locale.getDefault());
        for (int i = 0; i < texts.size(); i++) {
            final TextView item = (TextView) View.inflate(getContext(), R.layout.tag_text, null);
            final String str = texts.get(i).toString();
            final int idx = str.toLowerCase(Locale.getDefault()).indexOf(searchKey);
            if (idx != -1) {
                SpannableStringBuilder builder = new SpannableStringBuilder(texts.get(i));
                builder.setSpan(new android.text.style.StyleSpan(android.graphics.Typeface.BOLD),
                        idx,
                        idx + searchKey.length(),
                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                item.setText(builder);
            } else {
                item.setText(texts.get(i));
            }

            item.setOnClickListener(this);
            item.setOnLongClickListener(this);
            this.suggestionView.addView(item);
        }
    }

    @Override
    public void onFilter(String searchText, InlineAutocompleteEditText view) {
        // If the UrlInputFragment has already been hidden, don't bother with filtering. Because of the text
        // input architecture on Android it's possible for onFilter() to be called after we've already
        // hidden the Fragment, see the relevant bug for more background:
        // https://github.com/mozilla-mobile/focus-android/issues/441#issuecomment-293691141
        if (!isVisible()) {
            return;
        }
        autoCompleteInProgress = true;
        urlAutoCompleteFilter.onFilter(searchText, view);
        autoCompleteInProgress = false;
    }

    @Override
    public void onTextChange(String originalText, String autocompleteText) {
        if (autoCompleteInProgress) {
            return;
        }
        if (allowSuggestion) {
            UrlInputFragment.this.presenter.onInput(originalText, detectThrottle());
        }
        final int visibility = TextUtils.isEmpty(originalText) ? View.GONE : View.VISIBLE;
        UrlInputFragment.this.clearView.setVisibility(visibility);
    }

    private boolean detectThrottle() {
        long now = System.currentTimeMillis();
        boolean throttled = now - lastRequestTime < REQUEST_THROTTLE_THRESHOLD;
        lastRequestTime = now;
        return throttled;
    }
}

File: app/src/main/java/org/mozilla/focus/urlinput/UrlInputFragment.kt
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.urlinput

import android.content.res.Configuration
import android.graphics.Bitmap
import android.graphics.Color
import android.os.Bundle
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowInsets
import android.webkit.URLUtil
import android.widget.TextView
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.graphics.drawable.DrawableCompat
import androidx.core.graphics.drawable.toBitmap
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import dagger.Lazy
import kotlinx.android.synthetic.main.fragment_urlinput.input_container
import kotlinx.android.synthetic.main.fragment_urlinput.awesomeBar
import kotlinx.android.synthetic.main.fragment_urlinput.search_suggestion_block
import mozilla.components.browser.domains.autocomplete.ShippedDomainsProvider
import mozilla.components.ui.autocomplete.InlineAutocompleteEditText
import org.mozilla.focus.R
import org.mozilla.focus.navigation.ScreenNavigator
import org.mozilla.focus.search.SearchEngineManager
import org.mozilla.focus.telemetry.TelemetryWrapper
import org.mozilla.focus.telemetry.TelemetryWrapper.Extra_Value.AWESOMEBAR_TYPE_AUTOCOMPLETE
import org.mozilla.focus.telemetry.TelemetryWrapper.Extra_Value.AWESOMEBAR_TYPE_BOOKMARK
import org.mozilla.focus.telemetry.TelemetryWrapper.Extra_Value.AWESOMEBAR_TYPE_CLIPBOARD
import org.mozilla.focus.telemetry.TelemetryWrapper.Extra_Value.AWESOMEBAR_TYPE_HISTORY
import org.mozilla.focus.telemetry.TelemetryWrapper.Extra_Value.AWESOMEBAR_TYPE_MANUALCOMPLETE
import org.mozilla.focus.telemetry.TelemetryWrapper.Extra_Value.AWESOMEBAR_TYPE_SUGGESTION
import org.mozilla.focus.telemetry.TelemetryWrapper.Extra_Value.AWESOMEBAR_TYPE_TABTRAY
import org.mozilla.focus.utils.SearchUtils
import org.mozilla.focus.utils.SupportUtils
import org.mozilla.focus.utils.ViewUtils
import org.mozilla.focus.web.WebViewProvider
import org.mozilla.focus.widget.FlowLayout
import org.mozilla.rocket.awesomebar.BookmarkSuggestionProvider
import org.mozilla.rocket.awesomebar.ClipboardSuggestionProvider
import org.mozilla.rocket.awesomebar.HistorySuggestionProvider
import org.mozilla.rocket.awesomebar.SessionSuggestionProvider
import org.mozilla.rocket.chrome.ChromeViewModel
import org.mozilla.rocket.chrome.ChromeViewModel.OpenUrlAction
import org.mozilla.rocket.content.appComponent
import org.mozilla.rocket.content.getActivityViewModel
import org.mozilla.rocket.tabs.SessionManager
import org.mozilla.rocket.tabs.TabsSessionProvider
import org.mozilla.rocket.urlinput.QuickSearch
import org.mozilla.rocket.urlinput.QuickSearchAdapter
import org.mozilla.rocket.urlinput.QuickSearchViewModel
import java.util.Locale
import javax.inject.Inject

/**
 * Fragment for displaying he URL input controls.
 */
class UrlInputFragment : Fragment(), UrlInputContract.View, View.OnClickListener,
        View.OnLongClickListener, ScreenNavigator.UrlInputScreen {

    @Inject
    lateinit var quickSearchViewModelCreator: Lazy<QuickSearchViewModel>
    @Inject
    lateinit var chromeViewModelCreator: Lazy<ChromeViewModel>

    private val autoCompleteProvider: ShippedDomainsProvider = ShippedDomainsProvider()
    private lateinit var presenter: UrlInputContract.Presenter
    private lateinit var chromeViewModel: ChromeViewModel

    private lateinit var urlView: InlineAutocompleteEditText
    private lateinit var suggestionView: FlowLayout
    private lateinit var clearView: View
    private lateinit var dismissView: View
    private lateinit var quickSearchRecyclerView: RecyclerView
    private lateinit var quickSearchView: ViewGroup
    private var lastRequestTime: Long = 0
    private var autoCompleteInProgress: Boolean = false
    private var allowSuggestion: Boolean = false
    private var isUserInput = true

    override fun onCreate(bundle: Bundle?) {
        appComponent().inject(this)
        super.onCreate(bundle)
        val userAgent = WebViewProvider.getUserAgentString(activity)
        this.presenter = UrlInputPresenter(SearchEngineManager.getInstance()
                .getDefaultSearchEngine(activity), userAgent)
        chromeViewModel = getActivityViewModel(chromeViewModelCreator)

        context?.let {
            autoCompleteProvider.initialize(it.applicationContext)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_urlinput, container, false)

        dismissView = view.findViewById(R.id.dismiss)
        dismissView.setOnClickListener(this)

        clearView = view.findViewById(R.id.clear)
        clearView.setOnClickListener(this)

        suggestionView = view.findViewById<View>(R.id.search_suggestion) as FlowLayout

        urlView = view.findViewById<View>(R.id.url_edit) as InlineAutocompleteEditText
        urlView.setOnTextChangeListener(::onTextChange)
        urlView.setOnCommitListener(::onCommit)
        urlView.onFocusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
            // Avoid showing keyboard again when returning to the previous page by back key.
            if (hasFocus) {
                ViewUtils.showKeyboard(urlView)
            } else {
                ViewUtils.hideKeyboard(urlView)
            }
        }

        urlView.setOnFilterListener(::onFilter)

        initByArguments()

        initQuickSearch(view)

        return view
    }

    private fun toAwesomeBarIcon(res: Int): Bitmap? {
        return AppCompatResources.getDrawable(requireContext(), res)?.also {
            DrawableCompat.setTint(it, Color.WHITE)
        }?.toBitmap()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val ctx = context ?: return
        val bookmarkRepo = chromeViewModel.bookmarkRepo
        val historyRepo = chromeViewModel.historyRepo
        val iconBookmark = toAwesomeBarIcon(R.drawable.ic_illustration_no_bookmarks)
        val iconHistory = toAwesomeBarIcon(R.drawable.history_empty)
        val iconTab = toAwesomeBarIcon(R.drawable.ic_current_tab)

        awesomeBar.setOnApplyWindowInsetsListener { v: View, insets: WindowInsets ->
            (v.layoutParams as ViewGroup.MarginLayoutParams).topMargin = insets.systemWindowInsetTop
            insets
        }
        awesomeBar.addProviders(
                HistorySuggestionProvider(iconHistory, historyRepo) {
                    onSuggestionClicked(it, AWESOMEBAR_TYPE_HISTORY)
                },
                BookmarkSuggestionProvider(iconBookmark, bookmarkRepo) {
                    onSuggestionClicked(it, AWESOMEBAR_TYPE_BOOKMARK)
                },
                SessionSuggestionProvider(
                    iconTab,
                    TabsSessionProvider.getOrThrow(activity)
                ) { sm: SessionManager, id: String ->
                    sm.switchToTab(id)
                    ScreenNavigator[context].raiseBrowserScreen(false)
                    val isInLandscape = isInLandscape()
                    TelemetryWrapper.urlBarEvent(
                        isUrl = true,
                        isSuggestion = true,
                        isInLandscape = isInLandscape,
                        type = AWESOMEBAR_TYPE_TABTRAY
                    )
                },
                ClipboardSuggestionProvider(ctx) {
                    onSuggestionClicked(it, AWESOMEBAR_TYPE_CLIPBOARD)
                }
        )

        awesomeBar.onInputStarted() // if something is in the clipboard, it'll be the first and show directly
        awesomeBar.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                recyclerView.requestFocus()
                ViewUtils.hideKeyboard(recyclerView)
            }
        })
    }

    private fun initQuickSearch(view: View) {
        quickSearchView = view.findViewById(R.id.quick_search_container)
        quickSearchRecyclerView = view.findViewById(R.id.quick_search_recycler_view)
        quickSearchRecyclerView.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        val quickSearchAdapter = QuickSearchAdapter(fun(quickSearch: QuickSearch) {
            if (TextUtils.isEmpty(urlView.text)) {
                openUrl(quickSearch.homeUrl)
            } else {
                openUrl(quickSearch.generateLink(urlView.originalText))
            }
            TelemetryWrapper.clickQuickSearchEngine(quickSearch.name)
        })
        quickSearchRecyclerView.adapter = quickSearchAdapter
        getActivityViewModel(quickSearchViewModelCreator).run {
            quickSearchObservable.observe(
                    viewLifecycleOwner,
                    Observer { quickSearchList ->
                        quickSearchAdapter.submitList(quickSearchList)
                    }
            )
        }
    }

    override fun onStart() {
        super.onStart()
        presenter.setView(this)
        urlView.requestFocus()
    }

    override fun onStop() {
        super.onStop()
        presenter.setView(null)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        updateUrlInputHeight()
    }

    private fun updateUrlInputHeight() {
        val urlInputHeight = input_container.resources.getDimensionPixelOffset(R.dimen.search_url_input_height)
        input_container.layoutParams = input_container.layoutParams.apply {
            height = urlInputHeight
        }
    }

    override fun onLongClick(view: View): Boolean {
        when (view.id) {
            R.id.suggestion_item -> {
                setUrlText((view as TextView).text)
                TelemetryWrapper.searchSuggestionLongClick()
                return true
            }
            else -> return false
        }
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.clear -> {
                urlView.setText("")
                urlView.requestFocus()
                awesomeBar.onInputCancelled()
                TelemetryWrapper.searchClear(isInLandscape())
            }
            R.id.dismiss -> {
                dismiss()
                TelemetryWrapper.searchDismiss(isInLandscape())
            }
            R.id.suggestion_item -> onSuggestionClicked((view as TextView).text, AWESOMEBAR_TYPE_SUGGESTION)
            else -> throw IllegalStateException("Unhandled view in onClick()")
        }
    }

    private fun isInLandscape(): Boolean {
        return resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE
    }

    private fun initByArguments() {
        val args = arguments
        if (args?.containsKey(ARGUMENT_URL) == true) {
            val url = args.getString(ARGUMENT_URL)
            urlView.setText(url)
            clearView.visibility = if (TextUtils.isEmpty(url)) View.GONE else View.VISIBLE
        }
        allowSuggestion = args?.getBoolean(ARGUMENT_ALLOW_SUGGESTION, true) ?: false
        if (args?.getBoolean(ARGUMENT_BOOLEAN_PRIVATE_MODE, false) == true) {
            urlView.imeOptions = urlView.imeOptions or ViewUtils.IME_FLAG_NO_PERSONALIZED_LEARNING
        }
    }

    private fun onSuggestionClicked(tag: CharSequence, type: String) {
        val isInLandscape = isInLandscape()
        search(tag.toString())
        TelemetryWrapper.urlBarEvent(SupportUtils.isUrl(tag.toString()), true, isInLandscape, type)
    }

    private fun dismiss() {
        // This method is called from animation callbacks. In the short time frame between the animation
        // starting and ending the activity can be paused. In this case this code can throw an
        // IllegalStateException because we already saved the state (of the activity / fragment) before
        // this transaction is committed. To avoid this we commit while allowing a state loss here.
        // We do not save any state in this fragment (It's getting destroyed) so this should not be a problem.
        chromeViewModel.dismissUrlInput.call()
    }

    private fun onCommit() {
        val isInLandscape = isInLandscape()
        val input = urlView.autocompleteResult?.let { result ->
            if (result.text.isEmpty() || !URLUtil.isValidUrl(urlView.text.toString())) {
                urlView.text.toString()
            } else {
                result.text
            }
        } ?: urlView.text.toString()
        search(input)
        val type = if (isUserInput) {
            AWESOMEBAR_TYPE_MANUALCOMPLETE
        } else {
            AWESOMEBAR_TYPE_AUTOCOMPLETE
        }
        TelemetryWrapper.urlBarEvent(SupportUtils.isUrl(input), false, isInLandscape, type)
    }

    private fun search(input: String) {
        if (!input.trim { it <= ' ' }.isEmpty()) {
            ViewUtils.hideKeyboard(urlView)

            val isUrl = SupportUtils.isUrl(input)

            val url = if (isUrl)
                SupportUtils.normalize(input)
            else
                SearchUtils.createSearchUrl(context, input)

            val isOpenInNewTab = openUrl(url)

            if (isOpenInNewTab) {
                TelemetryWrapper.addNewTabFromHome()
            }
        }
    }

    /**
     * @param url the URL to open
     * @return true if open URL in new tab.
     */
    private fun openUrl(url: String): Boolean {
        val args = arguments
        val openNewTab = if (args != null && args.containsKey(ARGUMENT_PARENT_FRAGMENT)) {
            ScreenNavigator.HOME_FRAGMENT_TAG == args.getString(ARGUMENT_PARENT_FRAGMENT)
        } else {
            false
        }
        chromeViewModel.openUrl.value = OpenUrlAction(url, withNewTab = openNewTab, isFromExternal = false)

        return openNewTab
    }

    override fun setUrlText(text: CharSequence?) {
        text?.let {
            this.urlView.setText(text)
            this.urlView.setSelection(text.length)
        }
    }

    override fun setSuggestions(texts: List<CharSequence>?) {
        this.suggestionView.removeAllViews()
        search_suggestion_block.visibility = View.GONE
        if (texts == null) {
            return
        }

        val searchKey = urlView.originalText.trim { it <= ' ' }.toLowerCase(Locale.getDefault())
        if (texts.isNotEmpty()) {
            search_suggestion_block.visibility = View.VISIBLE
        }
        for (i in texts.indices) {
            val item = View.inflate(context, R.layout.tag_text, null) as TextView
            val str = texts[i].toString()
            val idx = str.toLowerCase(Locale.getDefault()).indexOf(searchKey)
            if (idx != -1) {
                val builder = SpannableStringBuilder(texts[i])
                builder.setSpan(android.text.style.StyleSpan(android.graphics.Typeface.BOLD),
                        idx,
                        idx + searchKey.length,
                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                item.text = builder
            } else {
                item.text = texts[i]
            }

            item.setOnClickListener(this)
            item.setOnLongClickListener(this)
            this.suggestionView.addView(item)
        }
    }

    override fun setQuickSearchVisible(visible: Boolean) {
    }

    private fun onFilter(searchText: String) {
        // If the UrlInputFragment has already been hidden, don't bother with filtering. Because of the text
        // input architecture on Android it's possible for onFilter() to be called after we've already
        // hidden the Fragment, see the relevant bug for more background:
        // https://github.com/mozilla-mobile/focus-android/issues/441#issuecomment-293691141
        if (!isVisible) {
            return
        }
        autoCompleteInProgress = true
        autoCompleteProvider.getAutocompleteSuggestion(searchText)?.let { result ->
            isUserInput = result.text == urlView.text.toString() || result.totalItems == 0
            urlView.applyAutocompleteResult(InlineAutocompleteEditText.AutocompleteResult(result.text, result.source, result.totalItems) { result.url })
        } ?: run {
            isUserInput = true
            urlView.applyAutocompleteResult(InlineAutocompleteEditText.AutocompleteResult(searchText, "", 0))
        }
        autoCompleteInProgress = false
    }

    private fun onTextChange(
        originalText: String,
        @Suppress("UNUSED_PARAMETER") autocompleteText: String
    ) {
        if (autoCompleteInProgress) {
            return
        }
        if (allowSuggestion) {
            awesomeBar.onInputChanged(originalText)
            this@UrlInputFragment.presenter.onInput(originalText, detectThrottle())
        }
        val visibility = if (TextUtils.isEmpty(originalText)) View.GONE else View.VISIBLE
        this@UrlInputFragment.clearView.visibility = visibility
    }

    private fun detectThrottle(): Boolean {
        val now = System.currentTimeMillis()
        val throttled = now - lastRequestTime < REQUEST_THROTTLE_THRESHOLD
        lastRequestTime = now
        return throttled
    }

    companion object {

        private const val ARGUMENT_URL = "url"
        private const val ARGUMENT_PARENT_FRAGMENT = "parent_frag_tag"
        private const val ARGUMENT_ALLOW_SUGGESTION = "allow_suggestion"
        private const val ARGUMENT_BOOLEAN_PRIVATE_MODE = "boolean_private_mode"
        private const val REQUEST_THROTTLE_THRESHOLD = 300

        /**
         * Create a new UrlInputFragment and animate the url input view from the position/size of the
         * fake url bar view.
         */
        @JvmStatic
        fun create(url: String?, parentFragmentTag: String?, allowSuggestion: Boolean, privateMode: Boolean = false): UrlInputFragment {
            val arguments = Bundle()
            arguments.putString(ARGUMENT_URL, url)
            arguments.putString(ARGUMENT_PARENT_FRAGMENT, parentFragmentTag)
            arguments.putBoolean(ARGUMENT_ALLOW_SUGGESTION, allowSuggestion)
            arguments.putBoolean(ARGUMENT_BOOLEAN_PRIVATE_MODE, privateMode)

            val fragment = UrlInputFragment()
            fragment.arguments = arguments

            return fragment
        }
    }

    override fun getFragment() = this
}
